apps/core/secureaction.cpp

```diff
@@ -0,0 +1,254 @@
+
+        std::shared_ptr<ServiceWatcher> watcher = instance.lock();
+
+        if (!watcher) {
+            watcher = std::make_shared<ServiceWatcher>();
+
+            instance = watcher;
+        }
+
+        return watcher;
+    }
+
+    bool isAvailable() const { return m_available; }
+
+signals:
+    void availableChanged(bool available);
+
+private:
+    bool m_available = false;;
+};
+
+SecureActionAuthenticatorAdaptor::SecureActionAuthenticatorAdaptor(SecureAction *action)
+    : QDBusAbstractAdaptor(action)
+    , m_secureAction(action)
+{
+}
+
+void SecureActionAuthenticatorAdaptor::PermissionGranted(uint)
+{
+    if (m_secureAction->m_authenticating) {
+        QJSValue resolve = m_secureAction->m_resolve;
+
+        m_secureAction->m_authorized = true;
+        m_secureAction->m_authenticating = false;
+        m_secureAction->m_resolve = QJSValue();
+
+        resolve.call();
+    }
+}
+
+void SecureActionAuthenticatorAdaptor::Aborted()
+{
+    m_secureAction->m_authenticating = false;
+    m_secureAction->m_resolve = QJSValue();
+}
+
+SecureAction::SecureAction(QObject *parent)
+    : QObject(parent)
+    , m_serviceWatcher(ServiceWatcher::instance())
+    , m_replyPath(replyPath())
+    , m_available(m_serviceWatcher->isAvailable())
+{
+    connect(m_serviceWatcher.get(), &ServiceWatcher::availableChanged, this, &SecureAction::setAvailable);
+}
+
+SecureAction::~SecureAction()
+{
+    cancel();
+}
+
+bool SecureAction::isAvailable() const
+{
+    return m_available;
+}
+
+QString SecureAction::message() const
+{
+    return m_message;
+}
+
+void SecureAction::setMessage(const QString &message)
+{
+    if (m_message != message) {
+        m_message = message;
+
+        m_authorized = false;
+        cancel();
+
+        emit messageChanged();
+    }
+}
+
+void SecureAction::perform(const QJSValue &resolve)
+{
+    if (m_authorized) {
+        QJSValue(resolve).call();
+    } else if (m_available && !m_message.isEmpty()) {
+        m_resolve = resolve;
+
+        if (m_authenticating) {
+            return;
+        }
+
+        QDBusConnection systemBus = QDBusConnection::systemBus();
+
+        if (!m_registered) {
+            m_registered = systemBus.registerObject(m_replyPath, this);
+
+            if (!m_registered) {
+                qmlInfo(this) << "DBus adaptor registration error " << systemBus.lastError().name() << " " << systemBus.lastError().message();
+                return;
+            }
+        }
+
+        m_authenticating = true;
+
+        m_response.reset(new QDBusPendingCallWatcher(systemBus.asyncCall(createAuthenticatorCall(
+                    QStringLiteral("RequestPermission"),
+                    { QVariant::fromValue(QDBusObjectPath(m_replyPath)), m_message, QVariantMap(), uint(Method::AllAvailable) }))));
+
+        connect(m_response.data(), &QDBusPendingCallWatcher::finished, this, [this](QDBusPendingCallWatcher *watcher) {
+            watcher->deleteLater();
+            m_response.take();
```

Comment:
I assume this is the same pointer as watcher?

Reply:
Yes